scipy/stats/_stats.pyx

```diff
@@ -300,3 +301,165 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
     tau = ((tot - (v + u - t)) - 2. * exchanges_weight[0]
            ) / np.sqrt(tot - u) / np.sqrt(tot - v)
     return min(1., max(-1., tau))
+
+
+# FROM MGCPY: https://github.com/neurodata/mgcpy
+
+# Distance transforms used for MGC and Dcorr
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _dense_rank_data(np.ndarray[np.float_t, ndim=1] x):
+    u, v = np.unique(x, return_inverse=True)
+    return v + 1
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _rank_distance_matrix(np.ndarray[np.float_t, ndim=2] distx):
+    # faster than np.apply_along_axis
+    return np.hstack([_dense_rank_data(distx[:, i]).reshape(-1, 1) for i in range(distx.shape[0])])
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _center_distance_matrix(np.ndarray[np.float_t, ndim=2] distx,
+                              str global_corr="mgc", is_ranked=True):
+    cdef int n = distx.shape[0]
+    cdef np.ndarray rank_distx = np.zeros((<object> distx).shape)
+
+    if is_ranked:
+        rank_distx = _rank_distance_matrix(distx)
+
+    if global_corr == "rank":
+        distx = rank_distx.astype(np.float)
+
+    # 'mgc' distance transform (col-wise mean) - default
+    cdef np.ndarray exp_distx = np.repeat(((distx.mean(axis=0) * n) / (n-1)), n).reshape(-1, n).T
+
+    # center the distance matrix
+    cdef np.ndarray cent_distx = distx - exp_distx
+
+    if global_corr != "mantel" and global_corr != "biased":
+        np.fill_diagonal(cent_distx, 0)
+
+    return cent_distx, rank_distx
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _transform_distance_matrix(np.ndarray[np.float_t, ndim=2] distx,
+                                 np.ndarray[np.float_t, ndim=2] disty,
+                                 str global_corr="mgc", is_ranked=True):
+
+    if global_corr == "rank":
+        is_ranked = True
+
+    cent_distx, rank_distx = _center_distance_matrix(distx, global_corr, is_ranked)
+    cent_disty, rank_disty = _center_distance_matrix(disty, global_corr, is_ranked)
+
+    transform_dist = {"cent_distx": cent_distx, "cent_disty": cent_disty,
+                      "rank_distx": rank_distx, "rank_disty": rank_disty}
+
+    return transform_dist
+
+
+# MGC specific functions
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _local_covariance(np.ndarray[np.float_t, ndim=2] distx,
+                        np.ndarray[np.float_t, ndim=2] disty,
```

Comment:
are we tightly bound to this older buffer syntax instead of modern typed memoryviews? I know I've had a hard time transitioning some of our cython to the newer format, but for new code it is nice to try!

Reply:
I think so. I think that it's fine to use `memoryviews` for two of the more basic functions, but for many of them, I am doing a lot of matrix arithmetic operations with `ndarrays`. Converting between the two becomes inefficient and actually raises run time by about 10 seconds for unit tests. For larger arrays, this may correspond to hours longer. If you know any workarounds for this I would be happy to implement it!

Reply:
>Converting between the two becomes inefficient and actually raises run time by about 10 seconds for unit tests.

In case this isn't obvious, you should never have to convert from view->`ndarray`  since manipulating the memory view modifies the array's contents inplace. e.g.
```python
float [:,:] cov_xy_view = cov_xy

for k in range(nx - 1):
    for l in range(ny - 1):
        cov_xy_view[k+1, l+1] += (
            cov_xy_view[k+1, l] + cov_xy_view[k, l+1]
            - cov_xy_view[k, l])
   
# Continue using cov_xy as normal
cov_xy[0, :] = np.cumsum(cov_xy[0, :])
```

Using the memory views for function arguments (e.g. in `_dense_rank_data`) is not a good idea since you now have to two additional constructors `ndarray`->view->`ndarray` where before there were none.

Reply:
https://github.com/scipy/scipy/pull/10524/commits/01fe019ff13f4b1664e9d0b34886d997224f9652